src/Dom/Document.php

```diff
@@ -0,0 +1,1033 @@
+ * @package AMP
+ */
+
+namespace Amp\AmpWP\Dom;
+
+use DOMComment;
+use DOMDocument;
+use DOMElement;
+use DOMNode;
+use DOMText;
+use DOMXPath;
+
+/**
+ * Class Amp\AmpWP\Dom\Document.
+ *
+ * @since 1.5
+ *
+ * @property DOMXPath   $xpath XPath query object for this document.
+ * @property DOMElement $head  The document's <head> element.
+ * @property DOMElement $body  The document's <body> element.
+ *
+ * Abstract away some of the difficulties of working with PHP's DOMDocument.
+ */
+final class Document extends DOMDocument {
+
+	/**
+	 * AMP requires the HTML markup to be encoded in UTF-8.
+	 *
+	 * @var string
+	 */
+	const AMP_ENCODING = 'utf-8';
+
+	/**
+	 * Encoding identifier to use for an unknown encoding.
+	 *
+	 * "auto" is recognized by mb_convert_encoding() as a special value.
+	 *
+	 * @var string
+	 */
+	const UNKNOWN_ENCODING = 'auto';
+
+	/**
+	 * Encoding detection order in case we have to guess.
+	 *
+	 * This list of encoding detection order is just a wild guess and might need fine-tuning over time.
+	 * If the charset was not provided explicitly, we can really only guess, as the detection can
+	 * never be 100% accurate and reliable.
+	 *
+	 * @var string
+	 */
+	const ENCODING_DETECTION_ORDER = 'UTF-8, EUC-JP, eucJP-win, JIS, ISO-2022-JP, ISO-8859-15, ISO-8859-1, ASCII';
+
+	/**
+	 * Attribute prefix for AMP-bind data attributes.
+	 *
+	 * @var string
+	 */
+	const AMP_BIND_DATA_ATTR_PREFIX = 'data-amp-bind-';
+
+	/**
+	 * Regular expression pattern to match the http-equiv meta tag.
+	 *
+	 * @var string
+	 */
+	const HTTP_EQUIV_META_TAG_PATTERN = '/<meta [^>]*?\s*http-equiv=[^>]*?>[^<]*(?:<\/meta>)?/i';
+
+	/**
+	 * Regular expression pattern to match the charset meta tag.
+	 *
+	 * @var string
+	 */
+	const CHARSET_META_TAG_PATTERN = '/<meta [^>]*?\s*charset=[^>]*?>[^<]*(?:<\/meta>)?/i';
+
+	/**
+	 * Regular expression pattern to match the main HTML structural tags.
+	 *
+	 * @var string
+	 */
+	const HTML_STRUCTURE_PATTERN = '/(?:.*?(?<doctype><!doctype(?:\s+[^>]*)?>))?(?:(?<pre_html>.*?)(?<html_start><html(?:\s+[^>]*)?>))?(?:.*?(?<head><head(?:\s+[^>]*)?>.*?<\/head\s*>))?(?:.*?(?<body><body(?:\s+[^>]*)?>.*?<\/body\s*>))?.*?(?:(?:.*(?<html_end><\/html\s*>)|.*)(?<post_html>.*))/is';
+
+	/*
+	 * Regular expressions to fetch the individual structural tags.
+	 * These patterns were optimized to avoid extreme backtracking on large documents.
+	 */
+	const HTML_STRUCTURE_DOCTYPE_PATTERN = '/^[^<]*<!doctype(?:\s+[^>]+)?>/i';
+	const HTML_STRUCTURE_HTML_START_TAG  = '/^[^<]*(?<html_start><html(?:\s+[^>]*)?>)/i';
+	const HTML_STRUCTURE_HTML_END_TAG    = '/(?:<\/html(?:\s+[^>]*)?>)[^<>]*$/i';
+	const HTML_STRUCTURE_HEAD_START_TAG  = '/^[^<]*(?:<head(?:\s+[^>]*)?>)/i';
+	const HTML_STRUCTURE_BODY_START_TAG  = '/^[^<]*(?:<body(?:\s+[^>]*)?>)/i';
+	const HTML_STRUCTURE_BODY_END_TAG    = '/(?:<\/body(?:\s+[^>]*)?>)[^<>]*$/i';
+	const HTML_STRUCTURE_HEAD_TAG        = '/^(?:[^<]*(?:<head(?:\s+[^>]*)?>).*?<\/head(?:\s+[^>]*)?>)/is';
```

Comment:
These patterns aren't all accounting for the inclusion of HTML comments which are added when doing validation requests. This results in HTML documents being corrupted when validating. See https://github.com/ampproject/amp-wp/issues/4104.